x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/recurrence_scheduler.tsx

```diff
@@ -0,0 +1,750 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { invert, mapValues } from 'lodash';
+import React, { useState, useCallback, useEffect, useMemo } from 'react';
+import moment, { Moment } from 'moment';
+import { i18n } from '@kbn/i18n';
+import { euiStyled } from '@kbn/kibana-react-plugin/common';
+import {
+  EuiDatePicker,
+  EuiPanel,
+  EuiFormRow,
+  EuiSelect,
+  EuiButtonGroup,
+  EuiFieldNumber,
+  EuiFormControlLayout,
+  EuiFormControlLayoutDelimited,
+  EuiHorizontalRule,
+} from '@elastic/eui';
+import { i18nMonthDayDate } from '../../../lib/i18n_month_day_date';
+import { RRuleFrequency, RecurrenceSchedule } from '../../../../types';
+
+const ISO_WEEKDAYS = [1, 2, 3, 4, 5, 6, 7];
+
+interface ComponentOpts {
+  startDate: Moment | null;
+  endDate: Moment | null;
+  initialState: RecurrenceSchedule | null;
+  onChange: (schedule: RecurrenceSchedule) => void;
+}
+
+interface CustomFrequencyState {
+  freq: RRuleFrequency;
+  interval: number;
+  byweekday: string[];
+  bymonthday: number[];
+  bymonth: number[];
+}
+
+export const RecurrenceScheduler: React.FC<ComponentOpts> = ({
+  startDate,
+  endDate,
+  onChange,
+  initialState,
+}) => {
+  const [hasInitialized, setHasInitialized] = useState(false);
+  const [frequency, setFrequency] = useState<RRuleFrequency | 'CUSTOM'>(RRuleFrequency.DAILY);
+  const [recurrenceEnds, setRecurrenceEnds] = useState('never');
+
+  const [customFrequency, setCustomFrequency] = useState<CustomFrequencyState>({
+    freq: RRuleFrequency.DAILY,
+    interval: 1,
+    byweekday: [],
+    bymonthday: [],
+    bymonth: [],
+  });
+
+  const [recurrenceEndDate, setRecurrenceEndDate] = useState(endDate);
+  const [occurrences, setOccurrrences] = useState(1);
+
+  const disableDailyOption = useMemo(() => {
+    if (!startDate || !endDate) return false;
+    return Math.abs(startDate.diff(endDate, 'hours')) >= 24;
+  }, [startDate, endDate]);
+
+  useEffect(() => {
+    if (disableDailyOption && frequency === RRuleFrequency.DAILY)
+      setFrequency(RRuleFrequency.WEEKLY);
+  }, [disableDailyOption, frequency]);
+
+  useEffect(() => {
+    if (initialState && !hasInitialized) {
+      const isCustomFrequency =
+        initialState.interval > 1 || (initialState.byweekday ?? []).length > 1;
+      setFrequency(isCustomFrequency ? 'CUSTOM' : initialState.freq);
+      if (isCustomFrequency) {
+        setCustomFrequency(initialState as CustomFrequencyState);
+      }
+      if (initialState.until) {
+        setRecurrenceEnds('ondate');
+        setRecurrenceEndDate(initialState.until);
+      }
+      if (initialState.count) {
+        setRecurrenceEnds('afterx');
+        setOccurrrences(initialState.count);
+      }
```

Comment:
a lot of state changes here, I am wondering if using a `useReducer` will make thing cleaner

Reply:
I disagree. There's 6 possible states that need to be hydrated on init and then this effect never gets called again. Every other interaction with these states happens individually, there's no other part of the code that sets multiple state values at once.

Reply:
what I was trying to express here, is that we have `compiledRecurrenceSchedule` which is under a `useMemo` and most of its dependency are under these different states and then we have `compiledRecurrenceSchedule` in the dependency of a `useEffect` which will call the onChange multiple time until all the state are set or initialize. I am wrong with my thinking here.